### PR TITLE
feat: Initial support for kafka timestamp pushdown.

### LINKED ISSUE #6361: batch: Implement kafka timestamp push down.
As in this doc: https://github.com/risingwavelabs/rfcs/pull/20
When the user limit offset by time range, we should use kafka's seek timestamp offset api to limit the start offset of each partition.

### PR DESCRIPTION
**Pull request by liurenjie1024**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Support limiting kafka query range through condition like `_rw_kafka_timestamp > '2022.-01-01 00:00:00+00:00'`.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* SQL commands, functions, and operators

### Release note

Now we can limit query kafka timestamp range through conditions like  `_rw_kafka_timestamp > '2022.-01-01 00:00:00+00:00'`.

## Refer to a related PR or issue link (optional)

Closes #6361

### CODE REVIEW

**Review by ZENOTME — APPROVED** [review]
cool!LGTM!

File: src/connector/src/source/kafka/enumerator/client.rs [hunk 1]
```diff
@@ -161,28 +164,45 @@ impl KafkaSplitEnumerator {
             }
             ret
         };
+        // println!("Watermark: {:?}", watermarks);
```
Comment (ZENOTME) [inline src/connector/src/source/kafka/enumerator/client.rs:167]: Why not use tracing here

File: src/frontend/src/handler/create_table.rs [hunk 2]
```diff
@@ -188,10 +188,11 @@ pub fn bind_sql_table_constraints(
     let mut columns_catalog = column_descs
         .into_iter()
         .map(|c| {
+            // All columns except `_row_id` or starts with `_rw` should be visible.
+            let is_hidden = c.name.starts_with("_rw");
```
Comment (tabVersion) [inline src/frontend/src/handler/create_table.rs:192]: > All columns except `_row_id` or starts with `_rw` should be visible.

is it a convention?

File: src/frontend/src/optimizer/plan_node/logical_source.rs [hunk 3]
```diff
@@ -114,13 +161,188 @@ impl ColPrunable for LogicalSource {
...
+///
+/// * `expr`: Expression to be consumed.
+/// * `range`: Original timestamp range, if `expr` can be recognized, we will update `range`.
+/// * `schema`: Input schema.
+///
+/// # Return Value
+///
+/// If `expr` can be recognized and consumed by this function, then we return `None`.
+/// Otherwise `expr` is returned.
+fn expr_to_kafka_timestamp_range(
```
Comment (tabVersion) [inline src/frontend/src/optimizer/plan_node/logical_source.rs:179]: I don't get why we still have a time range in the optimizer. Is not all splits that don't match time requirements are dismissed?

File: src/frontend/src/handler/create_source.rs [hunk 4]
```diff
@@ -83,23 +87,40 @@ pub async fn handle_create_source(
...
+
+    // Add hidden column `_rw_kafka_timestamp` to each message
+    if is_kafka && !is_materialized {
+        let kafka_timestamp_column = ColumnDesc {
+            data_type: DataType::Timestamptz,
+            column_id: ColumnId::new(column_descs.len() as i32),
+            name: KAFKA_TIMESTAMP_COLUMN_NAME.to_string(),
+            field_descs: vec![],
+            type_name: "".to_string(),
+        };
```
Comment (tabVersion) [inline src/frontend/src/handler/create_source.rs:109]: I don't see code that fills this field.

**Review by tabVersion — CHANGES_REQUESTED** [review]
will have further discussion about the logic and scalability of the code

hold the pr for now

File: src/connector/src/source/kafka/enumerator/client.rs (hunk 1 above)
Comment (liurenjie1024) [reply src/connector/src/source/kafka/enumerator/client.rs:167]: Just for debug, I will remove this.

File: src/frontend/src/handler/create_table.rs (hunk 2 above)
Comment (liurenjie1024) [reply src/frontend/src/handler/create_table.rs:192]: No, but currently we only have this hidden column, so I just want to leave it here.

File: src/frontend/src/handler/create_source.rs (hunk 4 above)
Comment (liurenjie1024) [reply src/frontend/src/handler/create_source.rs:109]: Which field?

File: src/frontend/src/optimizer/plan_node/logical_source.rs (hunk 3 above)
Comment (liurenjie1024) [reply src/frontend/src/optimizer/plan_node/logical_source.rs:179]: Timestamp range will apply to every split.

File: src/frontend/src/handler/create_source.rs (hunk 4 above)
Comment (ZENOTME) [reply src/frontend/src/handler/create_source.rs:109]: I think the funtion of this column(field) is that binder can use it to identify the '_rw_kafka_timestamp' in WHERE clause. And we don't need fill this field in row?

Comment (liurenjie1024) [reply src/frontend/src/handler/create_source.rs:109]: `field_desc` is used for struct type.

File: src/frontend/src/optimizer/plan_node/logical_source.rs (hunk 3 above)
Comment (ZENOTME) [reply src/frontend/src/optimizer/plan_node/logical_source.rs:179]: We will extract the timestamp literal in planner or optimizer. And use kafka_enumerator.list_splits_batch in plan_fragmet to covert timestamp to the according offset. 

BTW, this pr don't filter the invalid offset range. As we discuss in https://github.com/risingwavelabs/risingwave/pull/7112#discussion_r1059224681_, directly return empty vec if reader find the offset range is empty is good enough.

**Review by tabVersion — APPROVED** [review]

**Pull request merged**